### DISCUSSION TITLE
libmesh folder empty when installing MOOSE

### DISCUSSION THREAD

**Discussion by lodhia92** [idaholab/moose#23850] [Q&A Getting Started] [answered]
Hi, I am trying to install MOOSE on a virtual machine and am having some difficulties. I cannot save and install from my home directory (as suggested by the instructions), this is due to the settings on my work system. Instead, I am trying to install MOOSE and have all associated files in a mounted volume called /mydata

I am able to install Mamba and the MOOSE environment correctly (I think), since these steps in the instructions return no errors (see https://mooseframework.inl.gov/getting_started/installation/conda.html). However, When I get to the instruction step to run make -j 4, I get the following error:
```
(moose) lod021@brc-2303-749004:/mydata/projects/moose/test$ sudo make -j 4
/bin/sh: 1: /mydata/projects/moose/libmesh/installed/contrib/bin/libmesh-config: not found
/bin/sh: 1: /mydata/projects/moose/libmesh/installed/contrib/bin/libmesh-config: not found
/bin/sh: 1: /mydata/projects/moose/libmesh/installed/contrib/bin/libmesh-config: not found
/bin/sh: 1: /mydata/projects/moose/libmesh/installed/contrib/bin/libmesh-config: not found
/bin/sh: 1: /mydata/projects/moose/libmesh/installed/contrib/bin/libmesh-config: not found
/bin/sh: 1: /mydata/projects/moose/libmesh/installed/contrib/bin/libmesh-config: not found
/bin/sh: 1: /mydata/projects/moose/libmesh/installed/contrib/bin/libmesh-config: not found
/bin/sh: 1: /mydata/projects/moose/libmesh/installed/contrib/bin/libmesh-config: not found
/bin/sh: 1: /mydata/projects/moose/libmesh/installed/contrib/bin/libmesh-config: not found
/bin/sh: 1: /mydata/projects/moose/libmesh/installed/libtool: not found
/bin/sh: 1: /mydata/projects/moose/libmesh/installed/libtool: not found
/bin/sh: 1: /mydata/projects/moose/libmesh/installed/contrib/bin/libmesh-config: not found
/bin/sh: 1: /mydata/projects/moose/libmesh/installed/contrib/bin/libmesh-config: not found
/bin/sh: 1: /mydata/projects/moose/libmesh/installed/contrib/bin/libmesh-config: not found
make: python-config: No such file or directory
make: python-config: No such file or directory
Traceback (most recent call last):
  File "/mydata/projects/moose/framework/scripts/get_repo_revision.py", line 19, in <module>
    from packaging import version
ModuleNotFoundError: No module named 'packaging'
Checking if header needs updating: /mydata/projects/moose/framework/include/base/MooseRevision.h...

Failed to generate MooseRevision.h

Creating Unity /mydata/projects/moose/framework/build/unity_src/src_Unity.C
Symlinking MOOSE configure /mydata/projects/moose/framework/build/header_symlinks/MooseConfig.h
make: *** [/mydata/projects/moose/framework/moose.mk:325: /mydata/projects/moose/framework/include/base/MooseRevision.h] Error 1
make: *** Waiting for unfinished jobs....
```
I know the instructions say we shouldn't use sudo, however due to things not being located in my home directory as explained, I cannot avoid the use of sudo. I have checked and the folder /mydata/projects/moose/libmesh/ is empty. I've removed and repeated the earlier steps several times, however keep running into the same issue. Any help would be greatly appreciated!

**Comment by GiudGiud** [comment, marked as answer]
Hello

When you type sudo, my understanding is that you are running as an admin, and likely outside of the conda environment. 
So then it cant find libmesh.

You MUST not run as sudo. It's just a bad idea to be always running as sudo.
Can you fix the permissions in the folders instead? Using `chmod u+rw -R ./` in the /mydata and the home too while you're at it

Guillaume

**Reply by lodhia92** [reply]
Hi, thanks for the suggestion. I tried running `chmod u+rw -R ./` in /mydata, however this did not do anything and I still can't run any commands without the use of sudo.

When I try to do `mamba create --prefix=/mydata/anaconda3/envs/moose moose-tools moose-libmesh` without sudo to create the moose env, I get the following errors:

```
Looking for: ['moose-tools', 'moose-libmesh']

error    libmamba Could not open lockfile '/mydata/mambaforge3/pkgs/cache/cache.lock'
error    libmamba Could not open lockfile '/mydata/mambaforge3/pkgs/cache/cache.lock'
https://conda.software.inl.gov/public/linux-64                No change
https://conda.software.inl.gov/public/noarch                  No change
conda-forge/noarch                                  11.7MB @   4.4MB/s  2.7s
conda-forge/linux-64                                30.5MB @   5.1MB/s  6.2s
Transaction

  Prefix: /home/USER/.conda/envs/moose

  Updating specs:

   - moose-tools
   - moose-libmesh

```
I then removed the two files in `/mydata/mambaforge3/pkgs/cache` and ran the above again, however at the end got the following error:

```
Downloading and Extracting Packages

Preparing transaction: failed
Unable to create prefix directory '/mydata/anaconda3/envs/moose'.
Check that you have sufficient permissions.

CondaError: Unable to create prefix directory '/mydata/anaconda3/envs/moose'.
Check that you have sufficient permissions.
```

I'm not running commands using sudo as you suggested, however keep coming up against permission errors so am not sure how to proceed.

Thanks again for helping,

Bhavik

**Reply by milljm** [reply]
The issue is that somewhere along the line, `sudo` was used. And so `sudo` must now always be used when performing operations which require write access (such as `mamba <insert any command here>`). This is a woopsie mistake that should have been fixed immediately with `sudo rm -rf /mydata/mambaforge3`, and then reinstalling mambaforge without using sudo. 

>  I cannot save and install from my home directory (as suggested by the instructions), this is due to the settings on my work system

If this is true, this is a serious limitation imposed on by your administrator.

There is nothing preventing you from installing MOOSE to anywhere you have write access. So if you must install MOOSE outside of your home directory, that is fine. **Please attempt to reinstall mambaforge3 as suggested above.** The only constraints: you cannot use `sudo` when installing Mambaforge, or building or running MOOSE or building/running your MOOSE-based application.

If you _have_ to use `sudo`, then you must _always_ run as `root` instead: `sudo su -`. This is due to how environment functions are not inherited during the use of `sudo`. Example:

```bash
$ init_conda() { echo "hello" ; }
$ init_conda
hello
$ sudo init_conda
sudo: init_conda: command not found
```

There are functions being called when working with Conda. Like `mamba activate moose`. This will fail if you place `sudo` in front of that command.

**Reply by lodhia92** [reply]
Thanks for the response. I think I have sorted out the issue - since I am installing to a mounted volume, `sudo` had to be used when running commands in the volume itself, which kept leading to the errors. I've described this below in case any future users have the same issue.

I changed the ownership of the mounted volume by doing ` sudo chown $USER: /mydata` after which I did not need to use `sudo` any more. I was able to install `mamba` and then compile `moose` by following the instructions on the instructions page online but making sure all paths were correctly set to the folder I wanted them to be in.

Thanks for all of your help!

**Reply by GiudGiud** [reply]
Awesome. Thanks a lot for sharing the solution, might help us in the future!
